packages/core/src/async-value.ts

```diff
@@ -1,9 +1,12 @@
-export class AsyncValue<TValue> {
+export class AsyncValue<TValue> implements PromiseLike<TValue> {
+  public then: Promise<TValue>['then'];
+
   public constructor(
     public readonly promise: Promise<TValue>,
     public value?: TValue,
     public error?: Error
   ) {
+    this.then = promise.then.bind(promise) as Promise<TValue>['then'];
```

Comment:
This is necessary because TypeScript get tripped up with its inference chain by the `bind` call. I'm happy if someone has a better idea.